README
## Pepper8 - Transform pep8 or flake8 output to HTML

To install pepper8, simply use pip.

```
pip install pepper8
```

Or you can clone the latest master branch and build using setuptools

```
git clone git@github.com:myth/pepper8.git && python setup.py install
```

### Usage

pepper8 operates in two different input modes and two different output modes.
To read pep8 / flake8 status reports from file use

```
pepper8 -o report.html <filename>
```

If no output file is specified with `-o`, the HTML report is written to stdout, allowing
output redirection or piping

```
pepper8 <filename> | grep W301 > report.html
```

### Build statistics

When pepper8 is run using an output file `-o <filename>`, it will check its run environment
and print out available build statistics to stdout.

Currently, only TeamCity build statistics are supported with keys `pepper8warnings` and `pepper8errors`.

If you wish to enable full parsing of all PEP 8 / Flake8 errors as failures in TeamCity,
there already exists a package called [TeamCity-Messages](https://github.com/JetBrains/teamcity-messages)
which will be treated as build errors in TeamCity.

The TeamCity build statistics provided by pepper8 is only intended to provide statistical data for use in
custom build report graphs in addition to the HTML report.


CHANGELOG.rst
=========
Changelog
=========

1.0.3
-----
Fixed a bug causing missed statistics for last file.
Split the title summary stats into the Flake8 plugins.

1.0.2
-----
Changed links in HTML report to default to new window.
This prevents TeamCity to display blank page due to
loading of "unsafe scripts" inside the iframe.


1.0.1
-----
A few bugfixes


1.0.0
-----

Initial release



pepper8/main.py
# -*- coding: utf8 -*-
#
# Created by 'myth' on 10/19/15

import argparse
from os import fstat
from stat import S_ISFIFO, S_ISREG
from sys import stdin, stderr, exit, argv

from pepper8.generator import HtmlGenerator
from pepper8.parser import Parser

VERSION = '1.0.4'


def main(arguments=None):

    args = arguments or argv[1:]

    fileparser = None
    argparser = argparse.ArgumentParser(
        description='Convert pep8 or flake8 output to HTML',
        prog='pepper8',
        epilog='pepper8 accepts input either from stdin or from a filename argument.\n' +
               'Unless specified otherwise with -o OUTPUT_FILE, pepper8 outputs to stdout.'
    )
    argparser.add_argument(
        'filename',
        nargs='?',
        type=str,
        help='Path to file containing pep8 or flake8 results.'
    )
    argparser.add_argument(
        '-v',
        '--verbose',
        action='store_true',
        help='Enable verbose output (only if --output-file is specified)'
    )
    argparser.add_argument(
        '--version',
        action='store_true',
        help='Prints pepper8 version and exists'
    )
    argparser.add_argument(
        '-o',
        '--output-file',
        type=str,
        help='Outputs the HTML data to the specified file and enables the use of the --verbose option.'
    )

    # Fetch the provided arguments from sys.argv
    args = argparser.parse_args(args)
    if args.version:
        print('Pepper8 version %s' % VERSION)
        exit(0)

    if args.filename:
        try:
            f = open(args.filename)
            fileparser = Parser(f)
        except IOError as e:
            stderr.write('Could not open file: %s' % e)
            stderr.flush()
            exit(1)

    else:
        # We need to check if stdin is piped or read from file, since we dont want
        # stdin to hang at terminal input
        mode = fstat(0).st_mode

        if S_ISFIFO(mode) or S_ISREG(mode):
            fileparser = Parser(stdin)
        else:
            # stdin is terminal input at this point
            argparser.print_help()
            exit(0)

    # Generate the HTML report to output_file if not None, else print to stdout
    generator = HtmlGenerator(fileparser)
    generator.analyze(output_file=args.output_file)

if __name__ == '__main__':
    main()


pepper8/models.py
# -*- coding: utf8 -*-
#
# Created by 'myth' on 10/19/15


class ResultContainer(object):
    """
    Basic model that keeps the required fields for File and Package objects
    """

    def __init__(self):
        super(ResultContainer, self).__init__()
        self.lines = []
        self.violations = {}

    def add_error(self, code, line, char, description):
        """
        Registers an error for this container with code on line at char.
        :param code: The PEP 8 error code
        :param line: The line number of the reported error
        :param char: Line location of first offending character
        """

        if code not in self.violations:
            self.violations[code] = 0
        self.violations[code] += 1
        self.lines.append((code, line, char, description))


class FileResult(ResultContainer):
    """
    The FileResult class keeps the PEP 8 results for a single file
    """

    def __init__(self, path):
        super(FileResult, self).__init__()
        self.path = path


pepper8/parser.py
# -*- coding: utf8 -*-
#
# Created by 'myth' on 10/19/15

import re

POSITION = re.compile('^[\d]+$')
ERROR_CODE = re.compile('^[eEfFnNwW][0-9]{3,4}$')
FILEPATH = re.compile('[\S\s]+\.py$')


class Parser(object):
    """
    The Parser class is responsible for reading and parsing the lines from
    a PEP8 input source. This can be either piped through stdin or read from a file.
    """

    def __init__(self, data):
        """
        Constructs a Parser object.

        :param data: An input source of PEP 8 results. Can be either a file descriptor or stdin.
        """
        self.data = data

    def parse(self):
        """
        Reads all lines from the current data source and yields each FileResult objects
        """

        if self.data is None:
            raise ValueError('No input data provided, unable to parse')

        for line in self.data:
            parts = line.strip().split()
            try:
                path = parts[0]
                code = parts[1]
                path, line, char = path.split(':')[:3]

                if not re.match(POSITION, line):
                    continue
                if not re.match(POSITION, char):
                    continue
                if not re.match(ERROR_CODE, code):
                    continue
                if not re.match(FILEPATH, path):
                    continue

            # For parts mismatch
            except IndexError:
                continue
            # For unpack mismatch
            except ValueError:
                continue

            yield path, code, line, char, ' '.join(parts[2:])


pepper8/pepper8.py
# -*- coding: utf8 -*-
#
# Created by 'myth' on 10/19/15

import argparse
from os import fstat
from stat import S_ISFIFO, S_ISREG
from sys import stdin, stderr, exit, argv

from generator import HtmlGenerator
from parser import Parser


def main(arguments=None):

    args = arguments or argv[1:]

    fileparser = None
    argparser = argparse.ArgumentParser(
        description='Convert pep8 or flake8 output to HTML',
        prog='pepper8',
        epilog='pepper8 accepts input either from stdin or from a filename argument.\n' +
               'Unless specified otherwise with -o OUTPUT_FILE, pepper8 outputs to stdout.'
    )
    argparser.add_argument(
        'filename',
        nargs='?',
        type=str,
        help='Path to file containing pep8 or flake8 results.'
    )
    argparser.add_argument(
        '-v',
        '--verbose',
        action='store_true',
        help='Enable verbose output (only if --output-file is specified)'
    )
    argparser.add_argument(
        '-o',
        '--output-file',
        type=str,
        help='Outputs the HTML data to the specified file and enables the use of the --verbose option.'
    )

    # Fetch the provided arguments from sys.argv
    args = argparser.parse_args(args)

    if args.filename:
        try:
            f = open(args.filename)
            fileparser = Parser(f)
        except IOError as e:
            stderr.write('Could not open file: %s' % e)
            stderr.flush()
            exit(1)

    else:
        # We need to check if stdin is piped or read from file, since we dont want
        # stdin to hang at terminal input
        mode = fstat(0).st_mode

        if S_ISFIFO(mode) or S_ISREG(mode):
            fileparser = Parser(stdin)
        else:
            # stdin is terminal input at this point
            argparser.print_help()
            exit(0)

    # Generate the HTML report to output_file if not None, else print to stdout
    generator = HtmlGenerator(fileparser)
    generator.analyze(output_file=args.output_file)

if __name__ == '__main__':
    main()


pepper8/generator.py
# -*- coding: utf8 -*-
#
# Created by 'myth' on 10/19/15

import os
from sys import stdout, stderr

from jinja2 import Template

from pepper8.models import FileResult
from pepper8.parser import Parser


class HtmlGenerator(object):
    """
    The HTML Generator generates the HTML report based off of the data retrieved
    from the Parser object.
    """

    def __init__(self, parser):
        """
        Construct a HtmlGenerator object.

        :param parser: A pepper8 Parser object
        """

        if not isinstance(parser, Parser):
            raise ValueError('Argument <parser> must be an instance of Parser')

        super(HtmlGenerator, self).__init__()
        self.parser = parser
        self.files = {}
        self.total_errors = 0
        self.total_warnings = 0
        self.total_flakes = 0
        self.total_naming = 0
        self.total_complexity = 0
        self.violations = {}

    def analyze(self, output_file=None):
        """
        Analyzes the parsed results from Flake8 or PEP 8 output and creates FileResult instances
        :param output_file: If specified, output will be written to this file instead of stdout.
        """

        fr = None
        for path, code, line, char, desc in self.parser.parse():

            # Create a new FileResult and register it if we have changed to a new file
            if path not in self.files:
                # Update statistics
                if fr:
                    self.update_stats(fr)
                fr = FileResult(path)
                self.files[path] = fr

            # Add line to the FileResult
            fr.add_error(code, line, char, desc)

        # Add final FileResult to statistics, if any were parsed
        if fr:
            self.update_stats(fr)

        # Generate HTML file
        self.generate(output_file=output_file)

    def generate(self, output_file=None):
        """
        Generates an HTML file based on data from the Parser object and Jinja2 templates
        :param output_file: If specified, output will be written to this file instead of stdout.
        """

        fd = output_file

        # Write to stdout if we do not have a file to write to
        if not fd:
            fd = stdout
        else:
            try:
                fd = open(output_file, 'w')
            except IOError as e:
                stderr.write('Unable to open outputfile %s: %s' % (output_file, e))

        with open(os.path.join(os.path.dirname(__file__), 'templates/base.html')) as template:
            html = Template(template.read())

            # Write potential build messages to stdout if we are writing to HTML file
            # If dest is stdout and supposed to be piped or redirected, build messages like TeamCity's will
            # have no effect, since they require to read from stdin.
            if output_file:
                self.report_build_messages()

            # Write our rendered template to the file descriptor
            fd.write(
                html.render(
                    files=sorted(self.files.values(), key=lambda x: x.path),
                    total_warnings=self.total_warnings,
                    total_errors=self.total_errors,
                    total_naming=self.total_naming,
                    total_flakes=self.total_flakes,
                    total_complexity=self.total_complexity,
                    violations=sorted(
                        ((code, count) for code, count in self.violations.items()),
                        key=lambda x: x[1],
                        reverse=True
                    )
                )
            )

            # If file descriptor is stdout
            if not output_file:
                fd.flush()
            else:
                fd.close()

    def update_stats(self, file_result):
        """
        Reads the data from a FileResult and updates overall statistics
        :param file_result: A FileResult instance
        """

        for code, count in file_result.violations.items():
            if code not in self.violations:
                self.violations[code] = 0
            self.violations[code] += file_result.violations[code]

            if 'E' in code.upper():
                self.total_errors += file_result.violations[code]
            if 'W' in code.upper():
                self.total_warnings += file_result.violations[code]
            elif 'F' in code.upper():
                self.total_flakes += file_result.violations[code]
            elif 'N' in code.upper():
                self.total_naming += file_result.violations[code]
            elif 'C' in code.upper():
                self.total_complexity += file_result.violations[code]

    def report_build_messages(self):
        """
        Checks environment variables to see whether pepper8 is run under a build agent such as TeamCity
        and performs the adequate actions to report statistics.

        Will not perform any action if HTML output is written to OUTPUT_FILE and not stdout.
        Currently only supports TeamCity.

        :return: A list of build message strings destined for stdout
        """

        if os.getenv('TEAMCITY_VERSION'):
            tc_build_message_warning = "##teamcity[buildStatisticValue key='pepper8warnings' value='{}']\n"
            tc_build_message_error = "##teamcity[buildStatisticValue key='pepper8errors' value='{}']\n"

            stdout.write(tc_build_message_warning.format(self.total_warnings))
            stdout.write(tc_build_message_error.format(self.total_errors))
            stdout.flush()
